assimilation_code/modules/utilities/parse_args_mod.f90

```diff
@@ -554,7 +587,407 @@ subroutine get_next_arg(inline, startoff, argword, endoff)
+! We also need to treat empty fields so that we don't
+! collapse with the spaces and cause any column drifts. 
+! This serves as a wrapper for 'get_args_from_string'
+! Example: 
+! A;B;;;;C;; --> A B _EMPTY_ _EMPTY_ _EMPTY_ C _EMPTY_ _EMPTY_
+function normalize_delims(line, delim) result(out_line)
+
+character(len=*), intent(in)  :: line
+character,        intent(in)  :: delim
+
+character(len=MAX_FIELDS_LEN) :: out_line
+integer                       :: i, j, L, k, lee
+logical                       :: prev_is_delim
+
+! Start as with a delimiter 
+out_line      = ' '
+prev_is_delim = .true.
+
+j = 1
+L = len_trim(line)
+
+lee = len(EMPTY_ENTRY)
+
+! Go over the line 1 character at a time
+do i = 1, L
+   if (line(i:i) == char(13)) cycle
+   if (line(i:i) == delim) then
+      ! Found a delim
+      if (prev_is_delim) then
+         ! insert placeholder + 1 space
+         out_line(j:j+lee-1) = EMPTY_ENTRY
+         j = j+lee
+         out_line(j:j) = ' '
+
+         j = j+1
+      else
+         ! normal delimiter
+         out_line(j:j) = ' '
+         j = j+1
+      endif
+      prev_is_delim = .true.
+      if (j > MAX_FIELDS_LEN - 64) exit ! prevent overflow; 64 is a small cushion
+   else
+      out_line(j:j) = line(i:i) 
+
+      j = j+1
+      prev_is_delim = .false.
+      if (j > MAX_FIELDS_LEN - 64) exit
+   endif
+enddo
+
+! Trailing empty field: line ends with a delimiter (or several)
+if (L > 0 .and. line(L:L) == delim) then
+   out_line(j:j+lee-1) = EMPTY_ENTRY
+   j = j + lee
+endif
+
+! Trim right spaces
+k = j - 1
+do while (k >= 1 .and. out_line(k:k) == ' ')
+   k = k - 1
+enddo
+
+if (k < 1) then
+   out_line = ''
+else
+   out_line = out_line(1:k)
+endif
+
+end function normalize_delims
+
+
+!---------------------------------------------------
+! Find field index using cached header in csv_file_type.
+integer function csv_find_field(cf, key) result(idx)
+
+type(csv_file_type), intent(in) :: cf
+character(len=*),    intent(in) :: key
+
+integer           :: i
+character(len=64) :: field_name, field_key
+
+field_key = adjustl(trim(key))
+call to_upper(field_key)
+
+idx = -1
+
+do i = 1, cf%ncols
+   field_name = adjustl(trim(cf%fields(i)))
+   call to_upper(field_name)
+   if (field_name == field_key) then
+      idx = i
+      return
+   endif
+enddo
+
+end function csv_find_field
+
+
+!-------------------------------------------------
+! Detect the delimiter of a CSV file. One should expect a ','
+! but certain data files can come with a ';' so
+! let's take care of both cases
+function detect_delim(line) result(delim)
+
+character(len=*), intent(in) :: line
+
+character :: delim
+integer   :: i, nsemi, ncomma, L
+
+L = len_trim(line)
+
+nsemi  = 0
+ncomma = 0
+
+do i = 1, L
+   if (line(i:i) == ';') nsemi  = nsemi  + 1
+   if (line(i:i) == ',') ncomma = ncomma + 1
+enddo
+
+! Is there a better check? 
+if (nsemi >= ncomma) then
+   delim = ';'
+else
+  delim = ','
+endif
```

Comment:
this is making an educated guess because you could have the other character inside a field delimited by the other character, so the safer test is to let the caller tell you what it expects.  but it probably works most of the time if you don't have the info given to you